llarp/service/endpoint.cpp

```diff
@@ -808,12 +808,43 @@ namespace llarp
         handler(maybe);
         return true;
       }
-      auto path = PickRandomEstablishedPath();
-      if (path == nullptr)
-        return false;
       LogInfo(Name(), " looking up LNS name: ", name);
-      auto job = new LookupNameJob(this, GenTXID(), name, handler);
-      return job->SendRequestViaPath(path, m_router);
+      path::Path::UniqueEndpointSet_t paths;
+      ForEachPath([&](auto path) {
+        if (path->IsReady())
+        {
+          paths.insert(path);
+        }
+      });
+      // not enough paths
+      if (paths.size() < 3)
+      {
+        handler(std::nullopt);
+        return true;
+      }
+
+      auto maybeInvalidateCache = [handler, &cache, name](auto result) {
+        if (result)
+        {
+          cache.Put(name, *result);
+        }
+        else
+        {
+          cache.Remove(name);
+        }
+        handler(result);
+      };
+
+      auto resultHandler =
+          m_state->lnsTracker.MakeResultHandler(name, paths.size(), maybeInvalidateCache);
+
+      for (const auto& path : paths)
+      {
+        LogInfo(Name(), " lookup ", name, " from ", path->Endpoint());
+        auto job = new LookupNameJob(this, GenTXID(), name, resultHandler);
```

Comment:
ownership very unclear, I smell cursed shenanigans

Reply:
beware the abyss